sdk/java/pulumi/src/main/java/com/pulumi/serialization/internal/Converter.java

```diff
@@ -396,33 +396,45 @@ private Object tryConvertObjectInner(
+                    log.debug(String.format(
+                            "Expected type '%s' (annotated with '%s') to have a setter annotated with @%s(\"%s\"). " +
+                                    "Setter '%s' parameter named '%s' lacks @%s annotation, " +
+                                    "so the value is required, but there is no value to deserialize.",
+                            targetType.getTypeName(),
+                            CustomType.class.getTypeName(),
+                            CustomType.Setter.class.getTypeName(),
+                            name,
+                            setters.get(name),
+                            setters.get(name).getName(),
+                            Nullable.class.getTypeName()
+                    ));
+                }
                 try {
-                    if (!setters.containsKey(name)) {
-                        throw new IllegalArgumentException(String.format(
-                                "Expected type '%s' (annotated with '%s') to have a setter annotated with @%s(\"%s\")",
-                                targetType.getTypeName(),
-                                CustomType.class.getTypeName(),
-                                CustomType.Setter.class.getTypeName(),
-                                name
-                        ));
-                    }
-                    // validate null and @Nullable presence
-                    if (argument == null
-                            && !(extractSetterParameter(setters.get(name)).isAnnotationPresent(Nullable.class))) {
-                        log.debug(String.format(
-                                "Expected type '%s' (annotated with '%s') to have a setter annotated with @%s(\"%s\"). " +
-                                        "Setter '%s' parameter named '%s' lacks @%s annotation, " +
-                                        "so the value is required, but there is no value to deserialize.",
-                                targetType.getTypeName(),
-                                CustomType.class.getTypeName(),
-                                CustomType.Setter.class.getTypeName(),
-                                name,
-                                setters.get(name),
-                                setters.get(name).getName(),
-                                Nullable.class.getTypeName()
-                        ));
-                    }
-                    setters.get(name).invoke(builder, argument);
+                    var convertedArgument = tryConvertObjectInner(
+                            String.format("%s(%s)", targetType.getTypeName(), name),
+                            argument,
+                            TypeShape.extract(extractSetterParameter(setters.get(name)))
+                    );
+                    setters.get(name).invoke(builder, convertedArgument);
+                } catch (IllegalArgumentException e) {
```

Comment:
added info to IAE for debugging, previously got only "type mismatch error" in this case